qiskit/extensions/standard/barrier.py

```diff
@@ -41,19 +44,20 @@ def barrier(self, *qargs):
     For QuantumRegister, applies barrier to all the qubits in that register."""
     qubits = []
 
-    qargs = _convert_to_bits(qargs, [qbit for qreg in self.qregs for qbit in qreg])
-
     if not qargs:  # None
         for qreg in self.qregs:
             for j in range(qreg.size):
                 qubits.append((qreg, j))
 
     for qarg in qargs:
-        if isinstance(qarg, (QuantumRegister, list)):
-            if isinstance(qarg, QuantumRegister):
-                qubits.extend([(qarg, j) for j in range(qarg.size)])
-            else:
-                qubits.extend(qarg)
+        if isinstance(qarg, QuantumRegister):
+            qubits.extend([(qarg, j) for j in range(qarg.size)])
+        elif isinstance(qarg, list):
+            qubits.extend(qarg)
+        elif isinstance(qarg, range):
+            qubits.extend([i for i in qarg])
+        elif isinstance(qarg, slice):
+            qubits.extend([i for i in range(*qarg.indices(1000))])
```

Comment:
```suggestion
            qubits.extend(self.qubits[qarg])
```

Reply:
That moves the slice unfolding into qubits to this point, like is not happening in any other gate...

Reply:
bah.. `barrier` is so different that... why not...